src/coreclr/jit/assertionprop.cpp

```diff
@@ -4699,6 +4699,17 @@ GenTree* Compiler::optAssertionProp_Update(GenTree* newTree, GenTree* tree, Stat
             if (parent != nullptr)
             {
                 parent->ReplaceOperand(useEdge, newTree);
+                if ((parent->gtOper == GT_IND) && newTree->IsIconHandle())
```

Comment:
What has changed that requires modifying assertion prop?

Reply:
VN can fold the method table indir to a constant handle which this place then replaces indirs on such method table with. This causes an assert in `fgdiagnostic` to fire complaining about the indir missing the invariant and nonfaulting flags.

Reply:
But why couldn't this also happen without your changes? Is VN able to recognize something in your expansions but not the inlined IL intrinsics?

Reply:
Yes, so basically we had `IND(IND(knownObj))` and the inner `IND(knonwObj)` was folded by assertprop (with help of VN) to a `CLASS_HANDLE` so now we have `IND(CLASS_HANDLE)` but without proper flags (it's not really a correctness issue, more like annoying requirement from the fgdiagnostic)

Reply:
> Is VN able to recognize something in your expansions but not the inlined IL intrinsics?

I'd assume it's not able to recognize the field address + subtract magic there.

> But why couldn't this also happen without your changes?

I think this should reproduce with any indir over the method table returned from `gtNewMethodTableLookup`, seems like nothing does that today though.